### ISSUE TITLE
Empty list for google_compute_firewall source_ranges is implicitly converted to 0.0.0.0/0  

### ISSUE THREAD

**Issue by MaT1g3R** [terraform-providers/terraform-provider-google#4031]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment
* If an issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to "hashibot", a community member has claimed the issue already.

### Terraform Version
0.12.3

### Affected Resource(s)

* google_compute_firewall

### Terraform Configuration Files

```tf
resource "google_compute_firewall" "name" {
  name          = "name"
  direction     = "INGRESS"
  target_tags   = ["tag"]
  source_ranges = []
  allow {
    protocol = "tcp"
    ports    = ["22"]
  }
  priority = 1000
  network  = "default"
```

### Expected Behavior

An error from `terraform plan` to disallow empty list (suggest `127.0.0.0/32`)

### Actual Behavior

`0.0.0.0/0` is applied to the actual rule in Google Cloud

### Steps to Reproduce

1. `terraform plan`
1. `terraform apply`

**Comment by slevenick** [comment]
This default is applied by the firewall API itself, as documented here https://cloud.google.com/vpc/docs/using-firewalls

You may want to look into using `source_tags` as specifying a tag in `source_tags` will prevent the default `source_ranges` field from being populated with `0.0.0.0/0`

**Comment by MaT1g3R** [comment]
solved in #4032
